image_tools_py/test/test_showimage_cam2image.py.in

```diff
@@ -0,0 +1,114 @@
+
+    def generate_launch_description(self):
+        launch_description = LaunchDescription()
+        publisher_executable_args = ['-r', '1', '-s', '0', '-b', '-f', '5']
+        subscriber_executable_args = ['-r', '1', '-s', '0']
+
+        env = dict(os.environ)
+        env['PYTHONUNBUFFERED'] = '1'
+        env['OSPL_VERBOSITY'] = '8'  # 8 = OS_NONE
+        # bare minimum formatting for console output matching
+        env['RCUTILS_CONSOLE_OUTPUT_FORMAT'] = '{message}'
+        env['RMW_IMPLEMENTATION'] = r'@RMW_IMPLEMENTATION@'
+
+        # Launch the process that will receive the images.
+        # This is the process that gets to decide when to tear the launcher down.
+        # It will exit when the regex for receiving images is matched.
+        showimage_executable = r'@RCLPY_DEMO_SHOWIMAGE_EXE@'  # noqa: E501
+        showimage_name = 'test_showimage_py'
+
+        command = [showimage_executable]
+        command.extend(subscriber_executable_args)
+        showimage_process = ExecuteProcess(
+            cmd=command,
+            name=showimage_name,
+            env=env,
+            output='screen'
+        )
+        launch_description.add_action(showimage_process)
+
+        # Launch the process that will publish the images.
+        # This process will be exited when the launcher is torn down.
+        cam2image_executable = r'@RCLPY_DEMO_CAM2IMAGE_EXE@'  # noqa: E501
+        cam2image_name = 'test_cam2image_py'
+
+        command = [cam2image_executable]
+        command.extend(publisher_executable_args)
+        cam2image_process = ExecuteProcess(
+            cmd=command,
+            name=cam2image_name,
+            env=env,
+            output='screen'
+        )
+        launch_description.add_action(cam2image_process)
+
+        launch_description.add_action(
+            RegisterEventHandler(
+                OnProcessIO(
+                    on_stdout=self.append,
+                    on_stderr=self.append,
+                )
+            )
+        )
+
+        return launch_description
+
+    def append(self, process_io):
+        if 'cam2image' in process_io.process_name:
+            if 'Publishing image #' in process_io.text.decode():
+                self._saw_cam2image_output = True
+        elif 'showimage' in process_io.process_name:
+            if 'Received image #' in process_io.text.decode():
+                self._saw_showimage_output = True
+
+        if self._saw_cam2image_output and self._saw_showimage_output:
+            # We've seen all required arguments from the test, quit
+            return EmitEvent(event=Shutdown(reason='finished', due_to_sigint=False))
+
+        if time.time() - self._start_time > 10.0:
+            # We've waited more than 10 seconds with no output; raise an error
+            return EmitEvent(event=Shutdown(reason='timeout', due_to_sigint=False))
+
+    def test_reliable_qos(self):
+        self._ls.run()
+        self._ls.shutdown()
```

Comment:
@clalancette hmm if `run()` was left, then it must have shutdown the launch service already.

Reply:
Good point.  Fixed in 2335206